File: src-tauri/src/playback.rs
use rspotify::model::{
    AdditionalType::{Episode, Track},
    CurrentPlaybackContext, Device, RepeatState, TrackId,
};
use rspotify::prelude::*;
use tauri::{command, State};

use crate::SpotifyState;

#[command]
pub async fn get_playback_state(
    state: State<'_, SpotifyState>,
) -> Result<Option<CurrentPlaybackContext>, String> {
    state
        .get_spotify()
        .current_playback(None, Some([&Track, &Episode]))
        .await
        .map_err(|e| format!("{:#?}", e))
}

#[command]
pub async fn get_device_list(state: State<'_, SpotifyState>) -> Result<Vec<Device>, String> {
    state
        .get_spotify()
        .device()
        .await
        .map_err(|e| format!("{:#?}", e))
}

#[command]
pub async fn set_device(
    device_id: String,
    state: State<'_, SpotifyState>,
) -> Result<String, String> {
    state
        .get_spotify()
        .transfer_playback(&device_id, None)
        .await
        .map(|_| device_id)
        .map_err(|e| format!("{:#?}", e))
}

#[command]
pub async fn set_volume(volume_percent: u8, state: State<'_, SpotifyState>) -> Result<u8, String> {
    state
        .get_spotify()
        .volume(volume_percent, None)
        .await
        .map(|_| volume_percent)
        .map_err(|e| format!("{:#?}", e))
}

#[command]
pub async fn set_playing(play_state: bool, state: State<'_, SpotifyState>) -> Result<bool, String> {
    let result = if play_state {
        state.get_spotify().resume_playback(None, None).await
    } else {
        state.get_spotify().pause_playback(None).await
    };

    result.map(|_| play_state).map_err(|e| format!("{:#?}", e))
}

#[command]
pub async fn previous_track(state: State<'_, SpotifyState>) -> Result<(), String> {
    state
        .get_spotify()
        .previous_track(None)
        .await
        .map(|_| ())
        .map_err(|e| format!("{:#?}", e))
}

#[command]
pub async fn seek(progress: u32, state: State<'_, SpotifyState>) -> Result<u32, String> {
    let duration = chrono::Duration::milliseconds(progress.into());
    state
        .get_spotify()
        .seek_track(duration, None)
        .await
        .map(|_| progress)
        .map_err(|e| format!("{:#?}", e))
}

#[command]
pub async fn next_track(state: State<'_, SpotifyState>) -> Result<(), String> {
    state
        .get_spotify()
        .next_track(None)
        .await
        .map(|_| ())
        .map_err(|e| format!("{:#?}", e))
}

#[command]
pub async fn set_shuffle(
    shuffle_state: bool,
    state: State<'_, SpotifyState>,
) -> Result<bool, String> {
    state
        .get_spotify()
        .shuffle(shuffle_state, None)
        .await
        .map(|_| shuffle_state)
        .map_err(|e| format!("{:#?}", e))
}

#[command]
pub async fn set_repeat(
    repeat_state: RepeatState,
    state: State<'_, SpotifyState>,
) -> Result<RepeatState, String> {
    state
        .get_spotify()
        .repeat(repeat_state, None)
        .await
        .map(|_| repeat_state)
        .map_err(|e| format!("{:#?}", e))
}

#[command]
pub async fn check_liked(track_id: String, state: State<'_, SpotifyState>) -> Result<bool, String> {
    let limit = 50;
    let mut offset = 0;
    let spotify = state.get_spotify();

    while let Ok(page) = spotify
        .current_user_saved_tracks_manual(None, Some(limit), Some(offset))
        .await
    {
        for item in page.items {
            if item.track.is_local {
                return Err("Track is local, cannot save it.".to_string());
            }
            if item.track.id.unwrap().id() == track_id {
                return Ok(true);
            }
        }
        if page.next.is_none() {
            return Ok(false);
        }
        offset += limit;
    }
    Ok(false)
}

#[command]
pub async fn set_liked(
    track_id: String,
    liked_state: bool,
    state: State<'_, SpotifyState>,
) -> Result<bool, String> {
    let out = if liked_state {
        state
            .get_spotify()
            .current_user_saved_tracks_delete(vec![TrackId::from_id(&track_id).unwrap()])
            .await
    } else {
        state
            .get_spotify()
            .current_user_saved_tracks_add(vec![TrackId::from_id(&track_id).unwrap()])
            .await
    };
    out.map(|_| liked_state).map_err(|e| format!("{:#?}", e))
}

File: src/lib/playback.ts
import { asyncDerived, asyncWritable, derived, writable } from '@square/svelte-store';
import { invoke, window } from '@tauri-apps/api';
import ambient from '/ambient.svg?url';
import { get } from 'svelte/store';
import { clamp } from './utils';

type State = SpotifyApi.CurrentPlaybackResponse | null;
export type RepeatState = 'track' | 'context' | 'off';

export const authenticated = asyncWritable(
    [],
    () => invoke<boolean>('is_authenticated'),
    (newAuth) => invoke<boolean>(newAuth ? 'authenticate' : 'deauthenticate', { window }),
    { reloadable: true, initial: false }
);

export const loading = writable(false);
let lastState: State;

export const state = asyncDerived(
    authenticated,
    async ($auth) => {
        if ($auth) {
            loading.set(false);
            const state = await invoke<State>('get_playback_state');
            // Default case where a playback state is returned
            if (state) {
                lastState = state;
            }
            // Case for inactive state reached
            else if (lastState?.device?.id) {
                loading.set(true);
                // Set device to the current device to reset the inactivity timer
                invoke('set_device', { deviceId: lastState.device.id });
                // Set lastState to null so if the state is genuinenly disconnected,
                // we can return to null state.
                lastState = null;
            }
            return state ?? lastState;
        }
        return null;
    },
    { reloadable: true, initial: null }
);

export const devices = asyncDerived(
    authenticated,
    async ($auth) => (!$auth ? [] : await invoke<SpotifyApi.UserDevice[]>('get_device_list')),
    { reloadable: true, initial: [] }
);

const currentItem = derived(state, ($s) => $s?.item);
export const duration = derived(currentItem, ($i) => $i?.duration_ms ?? 0);
export const trackId = derived(currentItem, ($i) => $i?.id);
export const currentType = derived(state, ($s) => $s?.currently_playing_type);
export const images = derived([currentItem, currentType, trackId], ([$currentItem, $currentType]) => {
    if ($currentType == 'episode') return ($currentItem as SpotifyApi.EpisodeObject).images;
    else if ($currentType == 'track') return ($currentItem as SpotifyApi.TrackObjectFull).album.images;
    else return [{ url: ambient, width: 361, height: 361 }];
});

export const title = derived(currentItem, ($i) => $i?.name ?? '');
export const subheading = derived([currentItem, currentType, trackId], ([$currentItem, $currentType]) => {
    if ($currentType == 'episode') return ($currentItem as SpotifyApi.EpisodeObject)?.show?.name;
    else if ($currentType == 'track')
        return ($currentItem as SpotifyApi.TrackObjectFull)?.artists.map((a) => a.name).join(', ');
    else return '';
});

export const songLink = derived([currentType, trackId], ([$type, $trackId]) =>
    $type && $trackId ? `https://open.spotify.com/${$type}/${$trackId}` : undefined
);
export const deepLink = derived([currentType, trackId], ([$type, $trackId]) =>
    $type && $trackId ? `spotify://${$type}/${$trackId}` : undefined
);

export const activeDevice = asyncWritable(
    state,
    async ($state) => $state?.device?.id ?? null,
    (newDeviceId) => invoke('set_device', { deviceId: newDeviceId }),
    { reloadable: true, initial: null }
);

export const progress = asyncWritable(
    state,
    async ($state) => $state?.progress_ms ?? 0,
    (newProgress) => invoke<void>('seek', { progress: clamp(0, newProgress, get(duration)) }),
    { reloadable: true, initial: 0 }
);
export const displayedProgress = asyncWritable(
    [progress, duration],
    ([$progress]) => $progress,
    async (newValue, parents) => {
        const $duration = parents![1];
        if (newValue > $duration) loading.set(true);
        return clamp(0, newValue, $duration);
    }
);
export const playing = asyncWritable(
    state,
    async ($state) => $state?.is_playing ?? false,
    (newPlayState) => invoke<void>('set_playing', { playState: newPlayState }),
    { initial: false }
);
export const shuffle = asyncWritable(
    state,
    async ($state) => $state?.shuffle_state ?? false,
    (newShuffleState) => invoke<void>('set_shuffle', { shuffleState: newShuffleState }),
    { initial: false }
);
export const repeat = asyncWritable(
    state,
    async ($state) => ($state?.repeat_state ?? 'off') as RepeatState,
    (newRepeatState) => invoke<void>('set_repeat', { repeatState: newRepeatState }),
    { initial: 'off' as RepeatState }
);
export const liked = asyncWritable(
    trackId,
    async ($trackId) => (!$trackId ? false : await invoke<boolean>('check_liked', { trackId: $trackId })),
    async (likedState, $trackId) => {
        if ($trackId) return invoke<void>('set_liked', { trackId: $trackId, likedState });
    },
    { initial: false }
);
export const volume = asyncWritable(
    state,
    async ($state) => $state?.device?.volume_percent ?? 0,
    (newVolumePercent) => invoke<void>('set_volume', { volumePercent: clamp(0, newVolumePercent, 100) }),
    { initial: 0 }
);

export const disallows = derived(state, ($state) => {
    const disallowArray = ($state?.actions?.disallows as string[]) ?? { includes: () => true };
    return {
        playPause: $state?.is_playing ? disallowArray.includes('pausing') : disallowArray.includes('resuming'),
        seeking: disallowArray.includes('seeking'),
        skippingNext: disallowArray.includes('skipping_next'),
        skippingPrev: disallowArray.includes('skipping_prev'),
        changeVolume: $state?.device?.id === undefined,
        link: $state?.item?.id === undefined,
        togglingLike: $state?.item?.id === undefined,
        togglingRepeat:
            (disallowArray.includes('toggling_repeat_context') && disallowArray.includes('toggling_repeat_track')) ||
            ($state?.repeat_state === 'context' && disallowArray.includes('toggling_repeat_context')) ||
            ($state?.repeat_state === 'track' && disallowArray.includes('toggling_repeat_track')),
        togglingShuffle: disallowArray.includes('toggling_shuffle'),
        transferringPlayback: disallowArray.includes('transferring_playback'),
        interruptingPlayback: disallowArray.includes('interrupting_playback'),
    };
});
